lib/ukio/bookings/handlers/booking_creator.ex

```diff
@@ -1,13 +1,22 @@
 defmodule Ukio.Bookings.Handlers.BookingCreator do
   alias Ukio.Apartments
+  alias Ukio.Apartments.Handlers.AvailabilityChecker
 
   def create(
         %{"check_in" => check_in, "check_out" => check_out, "apartment_id" => apartment_id} =
           params
       ) do
-    with a <- Apartments.get_apartment!(apartment_id),
-         b <- generate_booking_data(a, check_in, check_out) do
-      Apartments.create_booking(b)
+    apartment = Apartments.get_apartment!(apartment_id)
+
+    if AvailabilityChecker.is_apartment_available?(
+         apartment_id,
+         check_in,
+         check_out
+       ) do
+      booking_data = generate_booking_data(apartment, check_in, check_out)
+      Apartments.create_booking(booking_data)
+    else
+      {:error, :apartment_not_available}
```

Comment:
Only create booking if the apartment is available, otherwise we return an error to be managed by the controller